polar-core/src/partial/constraints.rs

```diff
@@ -0,0 +1,501 @@
+        assert!(matches!(field.value(), Value::String(_)));
+
+        self.operations.push(op!(
+            Unify,
+            value.clone(),
+            term!(op!(Dot, self.variable_term(), field))
+        ));
+
+        let name = value.value().as_symbol().unwrap();
+        Term::new_temporary(Value::Partial(Constraints::new(name.clone())))
+    }
+
+    pub fn into_term(self) -> Term {
+        Term::new_temporary(Value::Partial(self))
+    }
+
+    /// Return the expression represented by this partial's constraints.
+    pub fn into_expression(self) -> Term {
+        Term::new_temporary(Value::Expression(Operation {
+            operator: Operator::And,
+            args: self
+                .operations
+                .into_iter()
+                .map(|op| Term::new_temporary(Value::Expression(op)))
+                .collect(),
+        }))
+    }
+
+    pub fn clone_with_name(&self, name: Symbol) -> Self {
+        let mut new = self.clone();
+        new.variable = name;
+        new
+    }
+
+    pub fn clone_with_operations(&self, operations: Vec<Operation>) -> Self {
+        let mut new = self.clone();
+        new.operations = operations;
+        new
+    }
+
+    pub fn name(&self) -> &Symbol {
+        &self.variable
+    }
+
+    fn variable_term(&self) -> Term {
+        Term::new_temporary(Value::Variable(sym!("_this")))
+    }
+}
+
+#[derive(Clone)]
+struct IsaConstraintCheck {
+    existing: Vec<Operation>,
+    proposed_tag: Option<Symbol>,
+    result: Option<bool>,
+    last_call_id: u64,
+}
+
+impl IsaConstraintCheck {
+    pub fn new(existing: Vec<Operation>, mut proposed: Operation) -> Self {
+        let right = proposed.args.pop().unwrap();
+        let proposed_tag = if let Value::Pattern(Pattern::Instance(instance)) = right.value() {
+            Some(instance.tag.clone())
+        } else {
+            None
+        };
+
+        Self {
+            existing,
+            proposed_tag,
+            result: None,
+            last_call_id: 0,
+        }
+    }
+
+    /// Check if constraint is compatible with proposed.
+    ///
+    /// Returns: None if compatible, QueryEvent::Done { false } if incompatible,
+    /// or QueryEvent to ask for compatibility.
+    fn check_constraint(
+        &mut self,
+        mut constraint: Operation,
+        counter: &Counter,
+    ) -> Option<QueryEvent> {
+        if constraint.operator != Operator::Isa {
+            return None;
+        }
+
+        let right = constraint.args.pop().unwrap();
+        if let Value::Pattern(Pattern::Instance(instance)) = right.value() {
+            let call_id = counter.next();
+            self.last_call_id = call_id;
+
+            // is_subclass check of instance tag against proposed
+            return Some(QueryEvent::ExternalIsSubclass {
+                call_id,
+                left_class_tag: self.proposed_tag.clone().unwrap(),
+                right_class_tag: instance.tag.clone(),
+            });
```

Comment:
Had a larger offline convo with @dhatch about this. tl;dr is that it's overly limiting to assume the only way two `isa` constraints are compatible is if the newer constraint class is a subclass of the older one.
Two examples that conflict:
1. The newer class is a superclass of the older one. We would still want this to match. Thinking of two rules that specialize on a superclass and a subclass. Even if the instance matches the subclass rule first, we would still want it to match the superclass rule (unless there is a `cut`). 
2. Multiple inheritance. E.g., in Python two unrelated classes could both be a superclass of a third class. Comparing an `isa` for these two unrelated classes would say they aren't compatible, even though the third class would match both.

Reply:
Thanks, will handle as follow up.